consensus/datasource/consensus_data_source.go

```diff
@@ -0,0 +1,78 @@
+package datasource
+
+import (
+	"context"
+	"errors"
+	syncmap "sync"
+	"sync/atomic"
+
+	"github.com/NethermindEth/juno/consensus/driver"
+	"github.com/NethermindEth/juno/consensus/p2p"
+	"github.com/NethermindEth/juno/consensus/proposer"
+	"github.com/NethermindEth/juno/consensus/types"
+	"github.com/NethermindEth/juno/core"
+	"github.com/NethermindEth/juno/sync"
+)
+
+const maxCommitHistory = 1024 // TODO: make this configurable
+
+type consensusDataSource[V types.Hashable[H], H types.Hash, A types.Addr] struct {
+	commitListener driver.CommitListener[V, H, A]
+	proposer       proposer.Proposer[V, H]
+	cache          syncmap.Map
+	latest         atomic.Uint64
+}
+
+func New[V types.Hashable[H], H types.Hash, A types.Addr](
+	commitListener driver.CommitListener[V, H, A],
+	proposer proposer.Proposer[V, H],
+	p2p p2p.P2P[V, H, A],
+) *consensusDataSource[V, H, A] {
+	return &consensusDataSource[V, H, A]{
+		commitListener: commitListener,
+		proposer:       proposer,
+		cache:          syncmap.Map{},
+		latest:         atomic.Uint64{},
+	}
+}
+
+func (c *consensusDataSource[V, H, A]) Run(ctx context.Context) error {
```

Comment:
Does this method returns an error to satisfy an interface? Otherwise I think we can return nothing

Reply:
Yes, currently the data source implementation for feeder gateway can return error.